### PR TITLE
fix tap migration renames with API

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Fixes https://github.com/Homebrew/brew/issues/17234

We don't load casks and formulae that used to exist in an external tap but have been moved and renamed to an API tap when using the API. Loading works correctly when the core formula or cask tap is tapped locally though.

This PR adds some logic to map tap migration renames to the API and check that when loading formulae and casks.

Todo:
- [x] Add tests
- [ ] Manual testing

### CODE REVIEW

**Review by Rylan12 — APPROVED** [review]
Makes sense to me! 👍

File: Library/Homebrew/cask/cask_loader.rb [hunk 1]
```diff
@@ -268,18 +268,28 @@ class FromAPILoader
...
       end
 
+      sig { params(ref: String).returns(T.nilable(String)) }
+      private_class_method def self.parse_token(ref)
+        return unless (token = ref[HOMEBREW_DEFAULT_TAP_CASK_REGEX, :token])
+        return token if Homebrew::API::Cask.all_casks.key?(token)
+
+        token if Homebrew::API::Cask.all_renames.key?(token)
+      end
+
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask_loader.rb:292]:
```suggestion
      sig { params(ref: String).returns(T.nilable(String)) }
      def self.parse_token(ref)
        return unless (token = ref[HOMEBREW_DEFAULT_TAP_CASK_REGEX, :token])
        return token if Homebrew::API::Cask.all_casks.key?(token)

        token if Homebrew::API::Cask.all_renames.key?(token)
      end
      private_class_method :parse_token

```

is what we more typically do here

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -941,6 +942,15 @@ def self.try_new(ref, from: T.unsafe(nil), warn: false)
...
     end
 
+    sig { params(ref: String).returns(T.nilable(String)) }
+    private_class_method def self.parse_name(ref)
+      return unless (name = ref[HOMEBREW_DEFAULT_TAP_FORMULA_REGEX, :name])
+      return name if Homebrew::API::Formula.all_formulae.key?(name)
+      return name if Homebrew::API::Formula.all_aliases.key?(name)
+
+      name if Homebrew::API::Formula.all_renames.key?(name)
+    end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:952]:
```suggestion
    sig { params(ref: String).returns(T.nilable(String)) }
    def self.parse_name(ref)
      return unless (name = ref[HOMEBREW_DEFAULT_TAP_FORMULA_REGEX, :name])
      return name if Homebrew::API::Formula.all_formulae.key?(name)
      return name if Homebrew::API::Formula.all_aliases.key?(name)

      name if Homebrew::API::Formula.all_renames.key?(name)
    end
    private_class_method :parse_name
```

**Review by MikeMcQuaid — APPROVED** [review]
Makes sense so far, thanks @apainintheneck!

**Comment by apainintheneck** [comment]
I've been thinking about this more and I just wanted to get some opinions on the order of execution for the formula and cask loaders. This PR changes things slightly for cask and formula short names. The question is how to handle ambiguous cask/formula short names.

### Before
- If the short name is a core cask/formula, it is evaluated to the core name even if it is ambiguous.
- If the short name is still ambiguous, it returns an error.

### After
- If the short name is a core cask/formula, it is evaluated to the core name even if it is ambiguous.
- If the short name has been migrated to a core cask/formula and we are using the API, it is evaluated to the core name even if it is ambiguous.
- If the short name is still ambiguous, it returns an error.

Does this sound reasonable? Would it be better to try to match the old behavior a bit more? That would probably involve modifying the name loader to call the API loader.

**Comment by MikeMcQuaid** [comment]
> Does this sound reasonable?

Sounds reasonable, yes, thanks @apainintheneck ❤️ 

If it helps: in the medium-term I think we should avoid having homebrew/core and homebrew/cask names ever overlapping at all. Given they are both "official" and "auto-tapped": it's unnecessarily confusing for our names to overlap. Many users don't know (or care) about the differences between formulae and casks.

That said: formulae are more widely used and should be the "default".

**Comment by apainintheneck** [comment]
The loading logic should be correct now. I tweaked things so that the warning message that the package has been renamed shows up correctly now. I still need to do some manual testing to make sure that the `brew migrate` command works correctly here. It might also be possible to add some tests for the migrator classes too.

File: Library/Homebrew/formulary.rb [hunk 3]
```diff
@@ -917,30 +917,41 @@ class FromAPILoader < FormulaLoader
     def self.try_new(ref, from: T.unsafe(nil), warn: false)
       return if Homebrew::EnvConfig.no_install_from_api?
       return unless ref.is_a?(String)
-      return unless (name = ref[HOMEBREW_DEFAULT_TAP_FORMULA_REGEX, :name])
-      if !Homebrew::API::Formula.all_formulae.key?(name) &&
-         !Homebrew::API::Formula.all_aliases.key?(name) &&
-         !Homebrew::API::Formula.all_renames.key?(name)
-        return
+
+      ref = if (name = parse_name(ref))
```
Comment (Rylan12) [inline Library/Homebrew/formulary.rb:921]: Thoughts on naming this something like:

```suggestion
      ref = if (name = parse_homebrew_core_name(ref))
```

Or something like that? That way it's clear that the "parsing" that's happening is only within the homebrew/core tap?

Comment (apainintheneck) [reply Library/Homebrew/formulary.rb:921]: Maybe `parse_core_name` here and `parse_core_token` in `Cask::CaskLoader`?

**Comment by apainintheneck** [comment]
Closing in favor of https://github.com/Homebrew/brew/pull/17599

**Comment by wenglaurie30** [comment]
fix-tap-migration-renames-with-api

**Comment by wenglaurie30** [comment]
fix-tap-migration-renames-with-api

**Comment by wenglaurie30** [comment]
e16bde13675120ea1e4e90bcece4953f8c20ea21

**Comment by wenglaurie30** [comment]
e16bde13675120ea1e4e90bcece4953f8c20ea21

**Pull request closed without merging**